Update fastly-rails

tyrauber opened the pull request:
fastly-rails hasn’t been updated in over 2 years. With the release of factly-ruby 2.3.0, that adds thread safety and concurrency, this gem needs to be updated as well.

Changes:

- update factly-ruby to 2.3.0
- factory_girl_rails is deprecated, migrate to factory_bot_rails
- remove old appraisals (deprecated ruby and rails versions and conflicts with bundler update to 2.0)
- test ruby 2.6.3 and rails 5.2.3
- fix test failures (get :show, params: {})
- fix minitest deprecations (asset_nil)
- Rails 5 before_action

tyrauber committed:
Update fastly-rails to use fastly 2.0

tyrauber committed:
Update fastly-rails
fastly-rails hasn’t been updated in over 2 years. With the release of factly-ruby 2.3.0, that adds thread safety and concurrency, this gem needs to be updated as well.

Changes:

- update factly-ruby to 2.3.0
- factory_girl_rails is deprecated, migrate to factory_bot_rails
- remove old appraisals (deprecated ruby and rails versions and conflicts with bundler update to 2.0)
- test ruby 2.6.3 and rails 5.2.3
- fix test failures (get :show, params: {})
- fix minitest deprecations (asset_nil)
- Rails 5 before_action

tyrauber commented:
We've been running this branch in production for the last two weeks, with a million+ uniques and 100K+ purges, without a single thread concurrency error.  This is ready to merge and addresses #79, #82 and #84.  Thanks a ton, @thommahoney, you're awesome.

thommahoney commented:
@tyrauber thanks for opening this PR. Do you know where rails publishes their actively supported versions? I found https://guides.rubyonrails.org/maintenance_policy.html

thommahoney commented on test/dummy/app/controllers/books_controller.rb:
```diff
@@ -1,6 +1,6 @@
-  before_filter :set_cache_control_headers, only: [:index, :show]
-  before_filter :find_book, :only => [:show, :edit, :update, :destroy]
+  before_action :set_cache_control_headers, only: [:index, :show]
+  before_action :find_book, :only => [:show, :edit, :update, :destroy]
```
do we need to update any docs?

tyrauber replied:
Updated docs and test suite.

thommahoney commented on test/fastly-rails_test.rb:
```diff
@@ -43,8 +43,8 @@
       assert_equal api_key, configuration.api_key
-      assert_equal user, configuration.user
-      assert_equal password, configuration.password
+      assert_nil user, configuration.user
```
this doesn't quite do what the original intent was. `assert_nil`'s second argurment is an optional message for when the assertion fails.

we should change this to be `assert_nil configuration.user` to keep the original intent of the test and avoid the minitest deprecation warning.

tyrauber replied:
Good catch! Unintentional and overlooked on my part. Reverted.

thommahoney reviewed (changes requested):
if you are able, could you address the two comments that I left? thank you again. 🙌

tyrauber commented:
> Do you know where rails publishes their actively supported versions? I found https://guides.rubyonrails.org/maintenance_policy.html

That's the closest you will find.  You can also look at the [rubygem releases](https://rubygems.org/gems/rails/versions) to see frequency of releases.  There was a security update pushed for 4.2.11.1 in March.  No one should be running anything less than that on production, and that is questionable, IMHO.

We could/should probably update the appraisals to drop support for 3.2.18, 4.0.5, and update 4.1.1 to 4.2.11.1 and  5.0 to 5.2.3.

tyrauber committed:
- Update docs and test re: before_action
- Ruby 2.6.3, 2.5.5, 2.4.6
- Rails 4.2.11.1, 5.2.3
- fix assert_equal in fastly-rails.test for user and password

tyrauber committed:
Drop 4.2.11.1 appraisal due to bundler conflict. Travis wants to use bundler 2.0 and < Rails 5 require < bundler 2.0, and > Rails 5 requires bundler 2.0.

tyrauber commented:
@thommahoney I updated travis to support Ruby 2.6.3, 2.5.5, and 2.4.6 Unfortunately, due to a incompatibility in bundler versions and rails versions, I was unable to figure out how to configure appraisal to support both Rails 5.2.3 and 4.2.11.1, as both require different versions of bundler.

satishkunisi commented:
This is awesome 🎉 I'm excited to finally get those thread errors resolved. Just a gentle bump to see if we can get this merged and released. Thank you @tyrauber @thommahoney !
